### PR TITLE
Add support for conformance groups

### PR DESCRIPTION
**Pull request by dthaler**
Per [latest ISA specification](https://htmlpreview.github.io/?https://raw.githubusercontent.com/ietf-wg-bpf/ebpf-docs/generated/draft-ietf-bpf-isa.html).

* Allow check.exe to specify conformance groups to include/exclude when verifying a program, defaulting to the set that the Linux kernel supports.
* For things other than check.exe, that use PREVAIL as a library, allow the platform structure to specify what conformance groups the platform supports.

### CODE REVIEW

File: src/main/check.cpp [hunk 1]
```diff
@@ -32,6 +32,44 @@ struct at_scope_exit
...
+    return result;
+}
+
+// Given a string containing comma-separated tokens, split them into a list of strings.
+static std::vector<std::string> get_string_vector(std::string list) {
+    std::vector<std::string> string_vector;
+    std::string element;
+    for (std::istringstream iss(list); std::getline(iss, element, ',');) {
+        string_vector.push_back(element);
+    }
```
Comment (elazarg) [inline src/main/check.cpp:69]:
```suggestion
    boost::split(string_vector, str, boost::is_any_of(","));
```

And include
```c++
#include <boost/algorithm/string.hpp>
```

(Though it's probably time we use `string_view`)

File: src/main/check.cpp [hunk 2]
```diff
@@ -88,6 +128,26 @@ int main(int argc, char** argv) {
...
+        } else {
+            std::cout << "Invalid group: " << group_name << std::endl;
+            return 1;
+        }
+    }
+    for (auto group_name : get_string_vector(exclude_groups)) {
+        if (auto group = _get_conformance_group_by_name(group_name)) {
+            platform.supported_conformance_groups &= ~(*group);
+        } else {
+            std::cout << "Invalid group: " << group_name << std::endl;
```
Comment (elazarg) [inline src/main/check.cpp:146]:
```suggestion
            std::cerr << "Invalid group: " << group_name << std::endl;
```

File: src/main/check.cpp [hunk 3]
```diff
@@ -88,6 +128,26 @@ int main(int argc, char** argv) {
...
     ebpf_verifier_options.allow_division_by_zero = !no_division_by_zero;
 
+    // Enable default conformance groups, which don't include callx or packet.
+    ebpf_platform_t platform = g_ebpf_platform_linux;
+    platform.supported_conformance_groups = bpf_conformance_groups_t::default_groups;
+    for (auto group_name : get_string_vector(include_groups)) {
+        if (auto group = _get_conformance_group_by_name(group_name)) {
+            platform.supported_conformance_groups |= *group;
+        } else {
+            std::cout << "Invalid group: " << group_name << std::endl;
```
Comment (elazarg) [inline src/main/check.cpp:138]:
```suggestion
            std::cerr << "Invalid group: " << group_name << std::endl;
```

**Review by elazarg — APPROVED** [review]
LGTM. I will merge after the minor comments are addressed.

File: src/main/check.cpp (hunk 2 above)
Comment (dthaler) [reply src/main/check.cpp]: Done

File: src/main/check.cpp (hunk 3 above)
Comment (dthaler) [reply src/main/check.cpp]: Done

File: src/main/check.cpp (hunk 1 above)
Comment (dthaler) [reply src/main/check.cpp]: Updated as suggested

**Comment by dthaler** [comment]
> LGTM. I will merge after the minor comments are addressed.

Thanks.  I think I've addressed the comments now.

**Pull request merged**